### ISSUE TITLE
Nested functions are not propagated properly

### ISSUE THREAD

**Issue by VBart** [nginx/njs#55]
The code below works fine:
```javascript
function foo(v) {
    function bar() {
        foo(0);
    }

    console.log(v);

    if (v === 1) {
        bar();
    }
}

foo(1);
```
It results as expected:
```
1
0
```

But if the whole code is wrapped into some other function:
```javascript
function baz() {
    function foo(v) {
        function bar() {
            foo(0);
        }

        console.log(v);

        if (v === 1) {
            bar();
        }
    }

    foo(1);
}

baz();
```
Then it doesn't work:
```
TypeError: object is not callable
    at baz (:1)
    at main (native)
```

**Comment by VBart** [comment]
One more example, that also doesn't work:
```javascript
var foo = function foo(v) {
    function bar() {
        foo(0);
    }

    console.log(v);

    if (v === 1) {
        bar(0);
    }
}

foo(1);
```

Results:
```
1
TypeError: object is not callable
    at bar (:3)
    at anonymous (:1)
    at main (native)
```

**Comment by drsm** [comment]
```js
>> var one = function() { function two() { console.log('two'); } console.log(one, two); two(); }
undefined
>> one()
[Function] [Function]
'two'
undefined
>> var one = function() { function two() { console.log(one, two); } console.log(one, two); two(); }
undefined
>> one();
[Function] undefined
TypeError: object is not callable
    at anonymous (:1)
    at main (native)
```

**Comment by xeioex** [comment]
Hi @VBart , thank you for the detailed report.

Please, try patch below

```diff
# HG changeset patch
# User Dmitry Volyntsev <xeioex@nginx.com>
# Date 1541093443 -10800
#      Thu Nov 01 20:30:43 2018 +0300
# Node ID d3d49880b6dc1dfe6780b6049e86b2f5eae4566e
# Parent  a3a0d5850b1a797eee0ebd24480acd47ab160275
Initializing closure values of a function.

This fixes #55 issue on Github.

diff --git a/njs/njs_function.c b/njs/njs_function.c
--- a/njs/njs_function.c
+++ b/njs/njs_function.c
@@ -399,7 +399,7 @@ njs_function_call(njs_vm_t *vm, njs_inde
     njs_ret_t              ret;
     nxt_uint_t             n, nesting;
     njs_frame_t            *frame;
-    njs_value_t            *value;
+    njs_value_t            *value, *init;
     njs_closure_t          *closure, **closures;
     njs_function_t         *function;
     njs_function_lambda_t  *lambda;
@@ -453,20 +453,19 @@ njs_function_call(njs_vm_t *vm, njs_inde
 
         if (size != 0) {
             closure = nxt_mem_cache_align(vm->mem_cache_pool,
-                                          sizeof(njs_value_t), size);
+                                          sizeof(njs_value_t), size + 1);
             if (nxt_slow_path(closure == NULL)) {
                 njs_memory_error(vm);
                 return NXT_ERROR;
             }
 
-            /* TODO: copy initialzed values. */
-
-            size -= sizeof(njs_value_t);
             closure->u.count = 0;
             value = closure->values;
 
+            init = lambda->closure_scope;
+
             do {
-                *value++ = njs_value_void;
+                *value++ = *init++;
                 size -= sizeof(njs_value_t);
             } while (size != 0);
         }
diff --git a/njs/njs_function.h b/njs/njs_function.h
--- a/njs/njs_function.h
+++ b/njs/njs_function.h
@@ -34,6 +34,7 @@ struct njs_function_lambda_s {
 
     /* Initial values of local scope. */
     njs_value_t                    *local_scope;
+    njs_value_t                    *closure_scope;
 
     union {
         u_char                     *start;
diff --git a/njs/njs_generator.c b/njs/njs_generator.c
--- a/njs/njs_generator.c
+++ b/njs/njs_generator.c
@@ -2047,11 +2047,13 @@ njs_generate_function_scope(njs_vm_t *vm
 
         if (closure != NULL) {
             lambda->block_closures = 1;
-            size = (1 + closure->items) * sizeof(njs_value_t);
+            lambda->closure_scope = closure->start;
+            size = closure->items * sizeof(njs_value_t);
         }
 
+        lambda->closure_size = size;
+
         lambda->nesting = node->scope->nesting;
-        lambda->closure_size = size;
         lambda->arguments_object = parser->arguments_object;
 
         lambda->local_size = parser->scope_size;
diff --git a/njs/test/njs_unit_test.c b/njs/test/njs_unit_test.c
--- a/njs/test/njs_unit_test.c
+++ b/njs/test/njs_unit_test.c
@@ -5287,6 +5287,49 @@ static njs_unit_test_t  njs_test[] =
     { nxt_string("function f() { var a = f2(); }"),
       nxt_string("ReferenceError: \"f2\" is not defined in 1") },
 
+    { nxt_string("(function(){ function f() {return f}; return f()})()"),
+      nxt_string("[object Function]") },
+
+    { nxt_string("var a = ''; "
+                 "function f(list) {"
+                 "    function add(v) {a+=v};"
+                 "    list.forEach(function(v) {add(v)});"
+                 "};"
+                 "f(['a', 'b', 'c']); a"),
+      nxt_string("abc") },
+
+    { nxt_string("(function(){ function f() {return f}; return f()})()"),
+      nxt_string("[object Function]") },
+
+    { nxt_string("var l = [];"
+                 "var f = function() { "
+                 "    function f2() { l.push(f); l.push(f2); }; "
+                 "    l.push(f); l.push(f2); "
+                 "    f2(); "
+                 "}; "
+                 "f(); "
+                 "l.every(function(v) {return typeof v == 'function'})"),
+      nxt_string("true") },
+
+    { nxt_string("var l = [];"
+                 "function baz() {"
+                 "  function foo(v) {"
+                 "     function bar() { foo(0); }"
+                 "     l.push(v);"
+                 "     if (v === 1) { bar(); }"
+                 "  }"
+                 "  foo(1);"
+                 "}; baz(); l"),
+      nxt_string("1,0") },
+
+    { nxt_string("var a = ''; "
+                 "function f(list) {"
+                 "    function add(v) {a+=v};"
+                 "    list.forEach(function(v) {add(v)});"
+                 "};"
+                 "f(['a', 'b', 'c']); a"),
+      nxt_string("abc") },
+
     /* Recursive fibonacci. */
 
     { nxt_string("function fibo(n) {"
```

**Comment by VBart** [comment]
It works. Thanks.
